compiler/rustc_mir_build/src/build/mod.rs

```diff
@@ -51,16 +51,20 @@ fn mir_build(tcx: TyCtxt<'_>, def: LocalDefId) -> Body<'_> {
             // of `mir_build`, so now we can steal it
             let thir = thir.steal();
 
-            tcx.ensure().check_match(def);
-            // this must run before MIR dump, because
-            // "not all control paths return a value" is reported here.
-            //
-            // maybe move the check to a MIR pass?
-            tcx.ensure().check_liveness(def);
-
-            match thir.body_type {
-                thir::BodyTy::Fn(fn_sig) => construct_fn(tcx, def, &thir, expr, fn_sig),
-                thir::BodyTy::Const(ty) => construct_const(tcx, def, &thir, expr, ty),
+            match tcx.check_match(def) {
```

Comment:
There is a call to `check_match` just above. Could we have only the first one?
There is no need to run lints on error MIR either, it's non-sensical anyway.

Reply:
I'm unfamiliar with the query system - does `tcx.ensure().check_match` need to be called again at this point for correctness or is that a mistake in the original code?

Reply:
`ensure` and `ensure_with_value` are weaker versions of the full call to `tcx.check_match`. They are do be used when the query's result is not used. Calling `tcx.check_match` call is a stronger statement, so all is well.